File: server/services/collections.ts
/**
 * Collaborative collections (US-007, tasklist 13 platform-infra).
 *
 * A collection is a user-curated group of entities (languages, cultures, battles,
 * …) gathered around a theme and shareable via URL. Persistence follows the
 * `contribution-service.ts` model — one JSON file per collection under
 * `data/collections/` — but the mutation/validation/ownership logic is factored
 * into **pure** functions (no fs, no clock) so CRUD is unit-tested directly and
 * the store is a thin filesystem wrapper over them.
 *
 * Entities are referenced by a **stable id** — the same deterministic
 * `cs:<node-type>:<linguascrape-id>` csid the shared-graph convergence work mints
 * (mirrors `graph-resolver.mintCsid` / tasklist 18 US-009), so a collection item
 * survives renames of denormalized display fields and lines up 1:1 with graph nodes.
 */

import crypto from "crypto";
import fs from "fs";
import path from "path";

// ============================================================================
// Types
// ============================================================================

/** A LinguaScrape entity reference — the same shape as the client `GraphEntityRef`. */
export interface CollectionEntityRef {
  /** Canonical node type, e.g. `"language"`, `"culture"`, `"battle"`. */
  type: string;
  /** LinguaScrape local id — the strong, stable signal. */
  id: string;
  /** Display name, denormalized so a collection renders without re-fetching. */
  name?: string;
  /** Optional region, denormalized for disambiguation/rendering. */
  region?: string;
}

/** One curated entity inside a collection. Keyed by its stable id. */
export interface CollectionItem {
  /** `cs:<type>:<id>` — the stable dedup/reference key. */
  stableId: string;
  ref: CollectionEntityRef;
  note?: string;
  /** ISO timestamp the item was added. */
  addedAt: string;
}

export type CollectionVisibility = "private" | "public";

export interface Collection {
  id: string;
  title: string;
  description?: string;
  /** Opaque owner id (no auth — supplied by the client, persisted per browser). */
  owner: string;
  visibility: CollectionVisibility;
  /** Unguessable token for URL sharing regardless of visibility. */
  shareToken: string;
  items: CollectionItem[];
  createdAt: string;
  updatedAt: string;
}

/** Public projection of a shared collection — never leaks the owner id. */
export interface CollectionShareView {
  id: string;
  title: string;
  description?: string;
  visibility: CollectionVisibility;
  items: CollectionItem[];
  itemCount: number;
  createdAt: string;
  updatedAt: string;
}

/** Metadata patch for an update (title/description/visibility). */
export interface CollectionPatch {
  title?: string;
  description?: string;
  visibility?: CollectionVisibility;
}

/** Fields accepted when creating a collection. */
export interface CollectionInput {
  title?: string;
  description?: string;
  visibility?: CollectionVisibility;
  items?: CollectionEntityRef[];
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

// ============================================================================
// Pure helpers (no fs, no clock)
// ============================================================================

/**
 * The stable reference key for an entity — mirrors `graph-resolver.mintCsid`
 * (`cs:<node-type>:<linguascrape-id>`). Kept local so this module stays free of
 * the graph-resolver's filesystem imports.
 */
export function stableEntityId(ref: Pick<CollectionEntityRef, "type" | "id">): string {
  return `cs:${ref.type}:${ref.id}`;
}

/** Coerce an arbitrary value to a valid visibility, defaulting to `private`. */
export function normalizeVisibility(value: unknown): CollectionVisibility {
  return value === "public" ? "public" : "private";
}

/** Validate an entity ref: `type` and `id` are required non-empty strings. */
export function validateEntityRef(ref: unknown): string[] {
  const errors: string[] = [];
  if (!ref || typeof ref !== "object") {
    errors.push("entity ref must be an object");
    return errors;
  }
  const r = ref as Record<string, unknown>;
  if (typeof r.type !== "string" || r.type.trim() === "") {
    errors.push("entity ref requires a non-empty 'type'");
  }
  if (typeof r.id !== "string" || r.id.trim() === "") {
    errors.push("entity ref requires a non-empty 'id'");
  }
  return errors;
}

/** Validate the payload for creating a collection. */
export function validateCollectionInput(input: Partial<CollectionInput>): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (typeof input.title !== "string" || input.title.trim() === "") {
    errors.push("title is required");
  } else if (input.title.length > 200) {
    errors.push("title must be 200 characters or fewer");
  }

  if (input.description !== undefined && typeof input.description !== "string") {
    errors.push("description must be a string");
  }

  if (
    input.visibility !== undefined &&
    input.visibility !== "private" &&
    input.visibility !== "public"
  ) {
    errors.push("visibility must be 'private' or 'public'");
  }

  if (input.items !== undefined) {
    if (!Array.isArray(input.items)) {
      errors.push("items must be an array");
    } else {
      input.items.forEach((item, i) => {
        for (const e of validateEntityRef(item)) errors.push(`items[${i}]: ${e}`);
      });
    }
  }

  return { valid: errors.length === 0, errors, warnings };
}

/** Normalize a ref down to its persisted, trimmed shape. */
function cleanRef(ref: CollectionEntityRef): CollectionEntityRef {
  const cleaned: CollectionEntityRef = { type: ref.type.trim(), id: ref.id.trim() };
  if (typeof ref.name === "string" && ref.name.trim() !== "") cleaned.name = ref.name.trim();
  if (typeof ref.region === "string" && ref.region.trim() !== "") cleaned.region = ref.region.trim();
  return cleaned;
}

/** Build a fresh item from a ref (assumes the ref has already been validated). */
function makeItem(ref: CollectionEntityRef, note: string | undefined, now: string): CollectionItem {
  const cleaned = cleanRef(ref);
  const item: CollectionItem = {
    stableId: stableEntityId(cleaned),
    ref: cleaned,
    addedAt: now,
  };
  if (typeof note === "string" && note.trim() !== "") item.note = note.trim();
  return item;
}

/**
 * Build a new collection from validated input. Ids/token/clock are injected so
 * the function stays pure and deterministic under test.
 */
export function createCollection(
  input: CollectionInput,
  ctx: { id: string; owner: string; shareToken: string; now: string },
): Collection {
  const items: CollectionItem[] = [];
  for (const ref of input.items ?? []) {
    const item = makeItem(ref, undefined, ctx.now);
    if (!items.some((i) => i.stableId === item.stableId)) items.push(item);
  }

  const collection: Collection = {
    id: ctx.id,
    title: input.title!.trim(),
    owner: ctx.owner,
    visibility: normalizeVisibility(input.visibility),
    shareToken: ctx.shareToken,
    items,
    createdAt: ctx.now,
    updatedAt: ctx.now,
  };
  const description = input.description?.trim();
  if (description) collection.description = description;
  return collection;
}

/** Apply a metadata patch, returning a new collection (does not mutate input). */
export function applyCollectionUpdate(
  collection: Collection,
  patch: CollectionPatch,
  now: string,
): Collection {
  const next: Collection = { ...collection, items: [...collection.items], updatedAt: now };
  if (patch.title !== undefined) next.title = patch.title.trim();
  if (patch.description !== undefined) {
    const description = patch.description.trim();
    if (description) next.description = description;
    else delete next.description;
  }
  if (patch.visibility !== undefined) next.visibility = normalizeVisibility(patch.visibility);
  return next;
}

/**
 * Add (or update) an entity in a collection. Deduped by stable id: re-adding an
 * existing entity refreshes its ref/note but keeps the original `addedAt`.
 */
export function addCollectionItem(
  collection: Collection,
  ref: CollectionEntityRef,
  note: string | undefined,
  now: string,
): Collection {
  const item = makeItem(ref, note, now);
  const existing = collection.items.find((i) => i.stableId === item.stableId);
  const items = existing
    ? collection.items.map((i) =>
        i.stableId === item.stableId ? { ...item, addedAt: i.addedAt } : i,
      )
    : [...collection.items, item];
  return { ...collection, items, updatedAt: now };
}

/** Remove an entity by stable id. Returns a new collection (no-op if absent). */
export function removeCollectionItem(
  collection: Collection,
  stableId: string,
  now: string,
): Collection {
  const items = collection.items.filter((i) => i.stableId !== stableId);
  if (items.length === collection.items.length) return collection;
  return { ...collection, items, updatedAt: now };
}

/** True when `owner` may read the collection (owner match or public). */
export function canView(collection: Collection, owner: string): boolean {
  return collection.owner === owner || collection.visibility === "public";
}

/** True when `owner` may mutate the collection (owner match only). */
export function canEdit(collection: Collection, owner: string): boolean {
  return collection.owner === owner;
}

/** Project a collection to its owner-free public share view. */
export function toShareView(collection: Collection): CollectionShareView {
  const view: CollectionShareView = {
    id: collection.id,
    title: collection.title,
    visibility: collection.visibility,
    items: collection.items,
    itemCount: collection.items.length,
    createdAt: collection.createdAt,
    updatedAt: collection.updatedAt,
  };
  if (collection.description) view.description = collection.description;
  return view;
}

// ============================================================================
// Store (filesystem persistence — thin wrapper over the pure functions)
// ============================================================================

/** Thrown when an owner attempts to mutate/view a collection they don't own. */
export class CollectionAccessError extends Error {
  constructor(message = "You do not have access to this collection") {
    super(message);
    this.name = "CollectionAccessError";
  }
}

export class CollectionStore {
  private dir: string;

  constructor(dataDir: string = "data/collections") {
    this.dir = path.resolve(dataDir);
    if (!fs.existsSync(this.dir)) fs.mkdirSync(this.dir, { recursive: true });
  }

  private filePath(id: string): string {
    return path.join(this.dir, `${id}.json`);
  }

  private writeFile(collection: Collection): void {
    fs.writeFileSync(this.filePath(collection.id), JSON.stringify(collection, null, 2), "utf-8");
  }

  private readAll(): Collection[] {
    const files = fs.readdirSync(this.dir).filter((f) => f.endsWith(".json"));
    return files.map((f) => JSON.parse(fs.readFileSync(path.join(this.dir, f), "utf-8")) as Collection);
  }

  /** Create + persist a collection owned by `owner`. */
  create(input: CollectionInput, owner: string): Collection {
    const collection = createCollection(input, {
      id: `col_${crypto.randomUUID()}`,
      owner,
      shareToken: crypto.randomBytes(12).toString("base64url"),
      now: new Date().toISOString(),
    });
    this.writeFile(collection);
    return collection;
  }

  /** All collections owned by `owner`, newest-updated first. */
  list(owner: string): Collection[] {
    return this.readAll()
      .filter((c) => c.owner === owner)
      .sort((a, b) => new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime());
  }

  /** Raw lookup by id (no access check). */
  get(id: string): Collection | null {
    const file = this.filePath(id);
    if (!fs.existsSync(file)) return null;
    return JSON.parse(fs.readFileSync(file, "utf-8")) as Collection;
  }

  /** Lookup by share token (for the public share view). */
  getByShareToken(token: string): Collection | null {
    if (!token) return null;
    return this.readAll().find((c) => c.shareToken === token) ?? null;
  }

  /** Update metadata; throws {@link CollectionAccessError} if `owner` can't edit. */
  update(id: string, patch: CollectionPatch, owner: string): Collection | null {
    const collection = this.get(id);
    if (!collection) return null;
    if (!canEdit(collection, owner)) throw new CollectionAccessError();
    const next = applyCollectionUpdate(collection, patch, new Date().toISOString());
    this.writeFile(next);
    return next;
  }

  /** Delete a collection; throws if `owner` can't edit. Returns false if absent. */
  remove(id: string, owner: string): boolean {
    const collection = this.get(id);
    if (!collection) return false;
    if (!canEdit(collection, owner)) throw new CollectionAccessError();
    fs.unlinkSync(this.filePath(id));
    return true;
  }

  /** Add an entity; throws if `owner` can't edit. */
  addItem(
    id: string,
    ref: CollectionEntityRef,
    note: string | undefined,
    owner: string,
  ): Collection | null {
    const collection = this.get(id);
    if (!collection) return null;
    if (!canEdit(collection, owner)) throw new CollectionAccessError();
    const next = addCollectionItem(collection, ref, note, new Date().toISOString());
    this.writeFile(next);
    return next;
  }

  /** Remove an entity by stable id; throws if `owner` can't edit. */
  removeItem(id: string, stableId: string, owner: string): Collection | null {
    const collection = this.get(id);
    if (!collection) return null;
    if (!canEdit(collection, owner)) throw new CollectionAccessError();
    const next = removeCollectionItem(collection, stableId, new Date().toISOString());
    this.writeFile(next);
    return next;
  }
}

File: services/api/src/pinakes/collab/collections.py
"""Collaborative collections — a user-curated, shareable group of entities.

The port of `server/services/collections.ts`. One JSON file per collection under
``data/runtime/collections``; items are keyed by the stable
``cs:<type>:<id>`` id so re-adding an entity refreshes it rather than
duplicating it.

Two rules are load-bearing and neither is obvious from the route surface:

* **Visibility governs reads, ownership governs writes.** A public collection is
  readable by anyone; only the owner may ever mutate one. The share token is a
  third, orthogonal thing — a capability that grants a read of the owner-free
  projection *regardless* of visibility, which is what makes "share a private
  collection by URL" work.
* **An unset optional is absent, not null.** `description` is deleted when a
  patch clears it, and never written as `null`, because the TypeScript reader
  distinguishes the two and `JSON.stringify` writes no key for `undefined`.

Pure functions take `id` / `shareToken` / `now` as arguments, exactly as the
TypeScript did, so every mutation rule is testable with no filesystem and no
clock; :class:`CollectionStore` is the thin wrapper that supplies them.
"""

from __future__ import annotations

import secrets
import uuid
from pathlib import Path
from typing import Any

from pinakes.collab.entities import (
    VISIBILITIES,
    Record,
    RecordStore,
    clean_ref,
    epoch_ms,
    normalize_visibility,
    stable_entity_id,
    validate_entity_ref,
)
from pinakes.contributions.store import ValidationResult, iso_now
from pinakes.paths import collections_dir

#: Longest accepted title. Bounds one JSON record; not a display constraint.
MAX_TITLE_LENGTH = 200

#: Bytes behind a share token. 12 → 16 base64url characters, which is what
#: `crypto.randomBytes(12).toString("base64url")` produces — the token is the
#: only thing protecting a shared private collection, so keep the width.
SHARE_TOKEN_BYTES = 12


class CollectionAccessError(Exception):
    """A non-owner tried to mutate a collection. The route maps this to 403."""

    def __init__(
        self, message: str = "You do not have access to this collection"
    ) -> None:
        super().__init__(message)
        self.message = message


# ── Validation (pure) ────────────────────────────────────────────────────────


def validate_collection_input(data: dict[str, Any]) -> ValidationResult:
    """The port of `validateCollectionInput`. Pure — no filesystem, no clock.

    Every optional is checked with ``"key" in data`` rather than a ``.get()``
    that cannot tell absent from ``null``: the TypeScript compares against
    ``undefined``, so an explicitly null `description` is a 400 while an omitted
    one is fine.
    """
    errors: list[str] = []

    title = data.get("title")
    if not isinstance(title, str) or title.strip() == "":
        errors.append("title is required")
    elif len(title) > MAX_TITLE_LENGTH:
        errors.append(f"title must be {MAX_TITLE_LENGTH} characters or fewer")

    if "description" in data and not isinstance(data["description"], str):
        errors.append("description must be a string")

    if "visibility" in data and data["visibility"] not in VISIBILITIES:
        errors.append("visibility must be 'private' or 'public'")

    if "items" in data:
        items = data["items"]
        if not isinstance(items, list):
            errors.append("items must be an array")
        else:
            for index, item in enumerate(items):
                for error in validate_entity_ref(item):
                    errors.append(f"items[{index}]: {error}")

    return ValidationResult(valid=not errors, errors=errors, warnings=[])


# ── Mutation (pure) ──────────────────────────────────────────────────────────


def make_item(ref: dict[str, Any], note: Any, now: str) -> Record:
    """One curated item from an already-validated ref."""
    cleaned = clean_ref(ref)
    item: Record = {
        "stableId": stable_entity_id(cleaned),
        "ref": cleaned,
        "addedAt": now,
    }
    if isinstance(note, str) and note.strip() != "":
        item["note"] = note.strip()
    return item


def create_collection(data: dict[str, Any], ctx: dict[str, str]) -> Record:
    """Build a collection from validated input. Ids/token/clock come in *ctx*."""
    items: list[Record] = []
    for ref in data.get("items") or []:
        item = make_item(ref, None, ctx["now"])
        if not any(existing["stableId"] == item["stableId"] for existing in items):
            items.append(item)

    collection: Record = {
        "id": ctx["id"],
        "title": str(data["title"]).strip(),
        "owner": ctx["owner"],
        "visibility": normalize_visibility(data.get("visibility")),
        "shareToken": ctx["shareToken"],
        "items": items,
        "createdAt": ctx["now"],
        "updatedAt": ctx["now"],
    }
    description = data.get("description")
    if isinstance(description, str) and description.strip() != "":
        collection["description"] = description.strip()
    return collection


def apply_collection_update(
    collection: Record, patch: dict[str, Any], now: str
) -> Record:
    """Apply a metadata patch, returning a new record (the input is not mutated).

    A patch key is only present when the route accepted its type, so the reads
    below need no re-validation — and a *blank* description is a deletion, which
    is why this is a `pop` and not an assignment of ``""``.
    """
    updated: Record = {
        **collection,
        "items": list(collection["items"]),
        "updatedAt": now,
    }
    if "title" in patch:
        updated["title"] = str(patch["title"]).strip()
    if "description" in patch:
        description = str(patch["description"]).strip()
        if description:
            updated["description"] = description
        else:
            updated.pop("description", None)
    if "visibility" in patch:
        updated["visibility"] = normalize_visibility(patch["visibility"])
    return updated


def add_collection_item(
    collection: Record, ref: dict[str, Any], note: Any, now: str
) -> Record:
    """Add or refresh an entity, deduped by stable id.

    Re-adding an entity updates its ref and note but **keeps the original
    `addedAt`** — the collection remembers when the entity first joined it, not
    when its display name was last refreshed.
    """
    item = make_item(ref, note, now)
    items = list(collection["items"])
    for index, existing in enumerate(items):
        if existing["stableId"] == item["stableId"]:
            items[index] = {**item, "addedAt": existing["addedAt"]}
            break
    else:
        items.append(item)
    return {**collection, "items": items, "updatedAt": now}


def remove_collection_item(collection: Record, stable_id: str, now: str) -> Record:
    """Remove an entity by stable id. Absent id ⇒ the record is returned as-is,
    `updatedAt` untouched — a no-op must not look like an edit."""
    items = [item for item in collection["items"] if item["stableId"] != stable_id]
    if len(items) == len(collection["items"]):
        return collection
    return {**collection, "items": items, "updatedAt": now}


def can_view(collection: Record, owner: str) -> bool:
    """Owner match, or the collection is public."""
    return collection.get("owner") == owner or collection.get("visibility") == "public"


def can_edit(collection: Record, owner: str) -> bool:
    """Owner match only. Public is a read grant, never a write one."""
    return collection.get("owner") == owner


def to_share_view(collection: Record) -> Record:
    """The owner-free public projection served for a share token.

    Dropping `owner` is the point — the token grants a read to someone who is
    not the owner, and the opaque id is the only thing resembling an identity in
    this surface. `shareToken` is dropped too: a viewer holds it already, and a
    re-share should be the owner's decision.
    """
    view: Record = {
        "id": collection["id"],
        "title": collection["title"],
        "visibility": collection["visibility"],
        "items": collection["items"],
        "itemCount": len(collection["items"]),
        "createdAt": collection["createdAt"],
        "updatedAt": collection["updatedAt"],
    }
    if collection.get("description"):
        view["description"] = collection["description"]
    return view


# ── The store ────────────────────────────────────────────────────────────────


class CollectionStore(RecordStore):
    """JSON-per-record collections rooted at one directory."""

    def create(self, data: dict[str, Any], owner: str) -> Record:
        """Create and persist a collection owned by *owner*."""
        collection = create_collection(
            data,
            {
                "id": f"col_{uuid.uuid4()}",
                "owner": owner,
                "shareToken": secrets.token_urlsafe(SHARE_TOKEN_BYTES),
                "now": iso_now(),
            },
        )
        self.write(collection)
        return collection

    def list_for_owner(self, owner: str) -> list[Record]:
        """Everything *owner* owns, newest-updated first.

        Named `list_for_owner` rather than `list`: a method called `list` shadows
        the builtin for every annotation after it in the class body, which strict
        mypy rejects outright (`server/CLAUDE.md`'s twin of the
        `ContributionStore.list` trap).
        """
        owned = [
            record for record in self.read_all() if record.get("owner") == owner
        ]
        owned.sort(key=lambda record: -epoch_ms(record.get("updatedAt")))
        return owned

    def get(self, collection_id: str) -> Record | None:
        """Raw lookup by id — **no access check**. Every caller applies its own."""
        return self.read(collection_id)

    def get_by_share_token(self, token: str) -> Record | None:
        """Lookup by share token. An empty token matches nothing, deliberately:
        a record whose `shareToken` failed to write must not become world-readable
        to a caller asking for ``/shared/``."""
        if not token:
            return None
        for record in self.read_all():
            if record.get("shareToken") == token:
                return record
        return None

    def update(
        self, collection_id: str, patch: dict[str, Any], owner: str
    ) -> Record | None:
        """Patch metadata. ``None`` when unknown; raises when *owner* may not edit."""
        collection = self.get(collection_id)
        if collection is None:
            return None
        if not can_edit(collection, owner):
            raise CollectionAccessError
        updated = apply_collection_update(collection, patch, iso_now())
        self.write(updated)
        return updated

    def remove(self, collection_id: str, owner: str) -> bool:
        """Delete a collection. ``False`` when unknown; raises on a non-owner."""
        collection = self.get(collection_id)
        if collection is None:
            return False
        if not can_edit(collection, owner):
            raise CollectionAccessError
        self.delete(collection_id)
        return True

    def add_item(
        self, collection_id: str, ref: dict[str, Any], note: Any, owner: str
    ) -> Record | None:
        """Add an entity. ``None`` when unknown; raises on a non-owner."""
        collection = self.get(collection_id)
        if collection is None:
            return None
        if not can_edit(collection, owner):
            raise CollectionAccessError
        updated = add_collection_item(collection, ref, note, iso_now())
        self.write(updated)
        return updated

    def remove_item(
        self, collection_id: str, stable_id: str, owner: str
    ) -> Record | None:
        """Remove an entity by stable id. ``None`` when unknown; raises on a
        non-owner."""
        collection = self.get(collection_id)
        if collection is None:
            return None
        if not can_edit(collection, owner):
            raise CollectionAccessError
        updated = remove_collection_item(collection, stable_id, iso_now())
        self.write(updated)
        return updated


def store(directory: Path | None = None) -> CollectionStore:
    """The collection store this process serves.

    Built per call from :func:`pinakes.paths.collections_dir`, which re-reads its
    environment override every time — the same no-singleton rule the contribution
    queue follows, and for the same two reasons: there is a second server reading
    this directory, and a test redirects it by setting one variable.
    """
    return CollectionStore(directory if directory is not None else collections_dir())
